### PR TITLE
Remove Virtual Terminal, Replace with Rich Strings

### LINKED ISSUE #2440: Refactor compiler messages to use rich strings and remove `VirtualTerminal`
The idea is to replace the `message`  and `summary` VirtualTerminal objects with rich strings like:

```
Foo bar baz <Red>This is Red</Red>
```

and then we will simply substitute `<Red>` and `</Red>` before printing.

I consider the VirtualTerminal as failed experiment.

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes #2440

- [x] Refactor Errors to return `String`
- [x] Introduce `Format` methods that produce rich strings
- [x] Remove `VirtualTerminal` and `VirtualString`
- [x] Remove References to `VirtualTerminal` and `VirtualString` (and refactor functions that depend on them)
- [x] Create mapping from rich string to actual formatting
- [x] Introduce `Formatter` Trait
- [x] Change signatures to `explain(Formatter): String, message(Formatter): String`

### CODE REVIEW

**Comment by jaschdoc** [comment]
It's not really an option that can be controlled by the end user rather than an environment specific thing so it would make sense for it to be moved into `Flix` as long as it can be injected into `Flix` upon creation s.t. the environment dictates which formatter to use.

**Comment by magnus-madsen** [comment]
> 
> 
> It's not really an option that can be controlled by the end user rather than an environment specific thing so it would make sense for it to be moved into `Flix` as long as it can be injected into `Flix` upon creation s.t. the environment dictates which formatter to use.

Let move the formatter to `Flix`. (but dont change the message and explain signatures- they should still just take an formatter- not the Flix object).

**Comment by jaschdoc** [comment]
I'm not quite sure which environments should use an AnsiTerminalFormatter and which ones don't. I imagine the packager is oblivious as to what formatter is used and the same goes for LanguageServer. SocketServer should just output the raw messages without formatting (no rich strings either), right?

**Comment by magnus-madsen** [comment]
> I'm not quite sure which environments should use an AnsiTerminalFormatter and which ones don't. I imagine the packager is oblivious as to what formatter is used and the same goes for LanguageServer. SocketServer should just output the raw messages without formatting (no rich strings either), right?

Language server should always just use PlainText. The rest should depend on the specific terminal in use.

Does that help? (It may not reflect how the code is currently).

**Comment by jaschdoc** [comment]
> Language server should always just use PlainText. The rest should depend on the specific terminal in use.

As far as I can tell that is what is written in the current code :)

File: main/src/ca/uwaterloo/flix/api/Flix.scala [hunk 1]
```diff
@@ -319,6 +324,18 @@ class Flix {
...
+    * Sets the output formatter used for this Flix instance.
+    */
+  def setFormatter(formatter: Formatter): Flix = {
+    if (formatter == null)
+      throw new IllegalArgumentException("'formatter' must be non-null.")
+    this.formatter = formatter
+    this
+  }
+
+  def getFormatter: Formatter = this.formatter
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/Flix.scala:337]: Add comment and move before setter.

File: main/src/ca/uwaterloo/flix/tools/Packager.scala [hunk 2]
```diff
@@ -167,20 +167,19 @@ object Packager {
...
-        errors.foreach(e => println(e.message.fmt))
+        errors.foreach(e => println(e.message(flix.getFormatter)))
     }
   }
 
   /**
     * Builds (compiles) the source files for the given project path `p`.
     */
-  def build(p: Path, o: Options, loadClasses: Boolean = true)(implicit tc: TerminalContext): Option[CompilationResult] = {
+  def build(p: Path, o: Options, flix: Flix = new Flix(), loadClasses: Boolean = true): Option[CompilationResult] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/tools/Packager.scala:177]: The flix argument should always go last and typically as an implicit parameter.

Where are you getting this object from?

File: main/src/ca/uwaterloo/flix/language/errors/WeederError.scala [hunk 3]
```diff
@@ -96,17 +101,23 @@ object WeederError {
...
+         |${code(loc2, "the second occurrence was here.")}
+         |""".stripMargin
     }
 
     def loc: SourceLocation = loc1
+
+    /**
+      * Returns a formatted string with helpful suggestions.
+      */
+    def explain(formatter: Formatter): Option[String] = None
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/WeederError.scala:120]: before def loc (which appears above)

**Review by magnus-madsen — COMMENTED** [review]
OK, I think its time we get this merged in, so we can iterate.

File: main/src/ca/uwaterloo/flix/tools/Packager.scala (hunk 2 above)
Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/tools/Packager.scala:177]: It's configured in Main

**Comment by magnus-madsen** [comment]
This was a big one. Its good to have it in!

**Pull request merged**